src/main/java/hudson/plugins/git/extensions/BrowserInference.java

```diff
@@ -0,0 +1,22 @@
+package hudson.plugins.git.extensions;
+
+import hudson.plugins.git.browser.GitRepositoryBrowser;
+import hudson.scm.RepositoryBrowsers;
+import org.apache.tools.ant.ExtensionPoint;
+
+public abstract class BrowserInference extends ExtensionPoint {
+
+    private GitRepositoryBrowser browser;
+
+    public BrowserInference(){
+        browser = null;
+    }
+
+    public BrowserInference(GitRepositoryBrowser browser){
+        this.browser = browser;
+    }
+
+    String inferRepositoryBrowser(){
+            return "no-browser";
+    }
```

Comment:
I might be misunderstanding Mark's suggestion, but from his words my understanding is exactly the opossite:
>  I assume that the extension point would need a method that would be passed the repository URL and would either return an instance of GitRepositoryBrowser or null. The null return would mean that the extension rejected the request to provide a browser, while if the GitRepositoryBrowser was returned, then that would be the inferred browser.

You have to receive a String/URL object and return the proper GitRepositoryBrowser. The place for the logic to calculate/detect the URL is there. Returning the GitRepositoryBrowser you will have the same effect as when you select a browser. At this point, have you run a Jenkins with the plugin installed and have you seen the current behaviour? So I can sure we are on the same page and we're understanding each other.